files/00_setup_freeipa.sh

```diff
@@ -27,25 +27,79 @@ else
     exit 254
 fi
 
+# Add a principal alias for the instance ID so folks can ssh in via
+# SSM Session Manager.  First check to make sure the alias does not
+# already exist.
+function add_principal {
+    instance_id=$(curl --silent \
+                       http://169.254.169.254/latest/meta-data/instance-id)
+
+    # domain and hostname are defined in the FreeIPA variables
+    # file that is sourced toward the top of this file.  Hence we
+    # can ignore the "undefined variable" warning from shellcheck.
+    #
+    # shellcheck disable=SC2154
+    if ipa host-show "$hostname" | \
+            grep "Principal alias" | \
+            grep --invert-match host/"$instance_id"."$domain"
+    then
+        ipa host-add-principal "$hostname" host/"$instance_id"."$domain"
+    fi
+}
+
 # Configure the host to be a FreeIPA client and join the domain.
-#
-# hostname is defined in the FreeIPA variables file that is sourced
-# toward the top of this file.  Hence we can ignore the "undefined
-# variable" warning from shellcheck.
-#
-# shellcheck disable=SC2154
-ipa-client-install --hostname="$hostname" \
-                   --mkhomedir \
-                   --no-ntp
+function install {
+    # hostname is defined in the FreeIPA variables file that is
+    # sourced toward the top of this file.  Hence we can ignore the
+    # "undefined variable" warning from shellcheck.
+    #
+    # shellcheck disable=SC2154
+    ipa-client-install --hostname="$hostname" \
+                       --mkhomedir \
+                       --no-ntp
 
-# Add a principal alias for the instance ID so folks can ssh in via
-# SSM Session Manager.
-#
-# domain is defined in the FreeIPA variables file that is sourced
-# toward the top of this file.  Hence we can ignore the "undefined
-# variable" warning from shellcheck.
-#
-# shellcheck disable=SC2154
-ipa host-add-principal \
-    "$hostname" \
-    host/"$(curl --silent http://169.254.169.254/latest/meta-data/instance-id)"."$domain"
+    add_principal
+}
+
+function enroll {
+    ipa-join
+    add_principal
+}
+
+function unenroll {
+    ipa-join --unenroll
+    # hostname is defined in the FreeIPA variables file that is
+    # sourced toward the top of this file.  Hence we can ignore the
+    # "undefined variable" warning from shellcheck.
+    #
+    # shellcheck disable=SC2154
+    ipa-rmkeytab -p "host/$hostname" -k /etc/krb5.keytab
```

Comment:
If `enroll` calls `ipa host-add-principal`, is there a reason why `unenroll` shouldn't call `ipa host-remove-principal`?

Reply:
Yes.  You may want to temporarily unenroll a host but _not_ remove it from the domain...the very reason why we decided to add this foo.

Reply:
Got it- just making sure.  Thanks!